### PR TITLE
Fix NRE with ConnectionStateChanged

### LINKED ISSUE #19448: Potential order manager (cleanup) regressions
I just got the following crash after leaving a skirmish lobby and immediately trying to join a new one:
```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object.
   at OpenRA.Mods.Common.Widgets.Logic.LobbyLogic.ConnectionStateChanged(OrderManager om, String password, NetworkConnection connection) in \OpenRA\OpenRA.Mods.Common\Widgets\Logic\Lobby\LobbyLogic.cs:line 77
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:line 665
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:line 832
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:line 885
   at OpenRA.Game.InitializeAndRun(String[] args) in \OpenRA\OpenRA.Game\Game.cs:line 280
   at OpenRA.Launcher.Program.Main(String[] args) in \OpenRA\OpenRA.Launcher\Program.cs:line 32
```

Apparently this seems to require some sort of timing, since I couldn't reproduce it right away. However I got to the following after leaving a skirmish lobby:
grafik
grafik
The dialogue is still displaying even when I'm not connecting to a lobby (in fact I was just leaving a lobby and returned to the main menu).

### PR DESCRIPTION
**Pull request by teinarss**
Was not able to reproduce the crash but looking at the stack trace this should fix it.

Closes: #19448

### CODE REVIEW

**Review by abcdefg30 — COMMENTED** [review]
Can you elaborate a bit more on why this fixes the issue? I don't quite get why we send a connection when it is already available via the order manager object. (Or are those different connections? If so why that?)
E.g. https://github.com/OpenRA/OpenRA/blob/b3b10729cda46763767598e337f4d208d65706ee/OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs#L25 still uses OrderManager.Connection while the LobbyLogic uses the connection parameter.

**Comment by teinarss** [comment]
> 
> 
> Can you elaborate a bit more on why this fixes the issue? I don't quite get why we send a connection when it is already available via the order manager object. (Or are those different connections? If so why that?)
> E.g.
> 
> https://github.com/OpenRA/OpenRA/blob/b3b10729cda46763767598e337f4d208d65706ee/OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs#L25
> still uses OrderManager.Connection while the LobbyLogic uses the connection parameter.

The NRE was caused by passing null here, (I think).
https://github.com/OpenRA/OpenRA/pull/19451/files#diff-730d2e90bb1459cb416c16468e2f5e4e771db77ea99af724fdc536bc34885af1R662 And the idea with sending the networkconnection in ConnectionStateChanged was that we could remove `ConnectionState` from `IConnection`, its only used for the networkconnection but i think some parts got lost when i pulled this code out from another PR.

File: OpenRA.Game/Game.cs [hunk 1]
```diff
@@ -653,10 +656,10 @@ static void LogicTick()
 		{
 			PerformDelayedActions();
 
-			if (OrderManager.Connection.ConnectionState != lastConnectionState)
+			if (connection != null && connection.ConnectionState != lastConnectionState)
```
Comment (abcdefg30) [inline OpenRA.Game/Game.cs]: What would be wrong with using `OrderManager.Connection != null && ...` here instead of a static variable?

Comment (pchote) [reply OpenRA.Game/Game.cs]: This sounds reasonable to me, except `OrderManager.Connection` can't be null so it simplifies even further. What do you think @teinarss?

Comment (teinarss) [reply OpenRA.Game/Game.cs]: We still need to check that the OrderManager.Connection is a NetworkConnection

Comment (pchote) [reply OpenRA.Game/Game.cs]: Would something like this work?
```csharp
if (OrderManager.Connection is NetworkConnection nc && nc.ConnectionState != lastConnectionState)
{
	lastConnectionState = nc.ConnectionState;
	ConnectionStateChanged(OrderManager, null, nc);
}
```

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Mods.Common/Widgets/Logic/DisconnectWatcherLogic.cs [hunk 2]
```diff
@@ -22,14 +22,19 @@ public DisconnectWatcherLogic(Widget widget, OrderManager orderManager)
 			var disconnected = false;
 			widget.Get<LogicTickerWidget>("DISCONNECT_WATCHER").OnTick = () =>
 			{
-				if (disconnected || orderManager.Connection.ConnectionState != ConnectionState.NotConnected)
+				var connection = orderManager.Connection as NetworkConnection;
+
+				if (connection == null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/DisconnectWatcherLogic.cs]: Worth reducing to
```csharp
if (!(orderManager.Connection is NetworkConnection connection))
	return;
```
?

**Review by pchote — DISMISSED** [review]
Code changes look sensible, but I also don't know how to reproduce the crash (have seen it a few times on bleed but can't trigger it on demand).

**Comment by teinarss** [comment]
Updated.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/72d542bfe70adb9c159c19566b0a158465017527...538d78294fc5fb97b5d9ca74477d0de45327f9d7)

**Pull request merged**